File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
from collections import deque

def solution(land):
    depth = len(land)
    oil_num = len(land[0])
    oils = [0 for i in range(oil_num)]
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    
    for y in range(depth):
        for x in range(oil_num):
            if land[y][x] == 0:
                continue
                
            qu = deque()        
            qu.append((y, x))
            land[y][x] = 0
            connect = set()
            amount = 0
            
            while qu:
                y1, x1 = qu.popleft()
                amount += 1
                connect.add(x1)
                
                for direction in directions:
                    dx, dy = direction
                    ny, nx = y1 + dy, x1 + dx
                    if 0 <= nx < oil_num and 0 <= ny < depth:
                        if land[ny][nx] == 1:
                            qu.append((ny, nx))
                            land[ny][nx] = 0
            
            for num in connect:
                oils[num] += amount
                
    return max(oils)

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;

public class Solution {
    public int solution(int[][] land) {
        int depth = land.length;
        int oilNum = land[0].length;
        int[] oils = new int[oilNum];
        int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        
        for (int y = 0; y < depth; y++) {
            for (int x = 0; x < oilNum; x++) {
                if (land[y][x] == 0) {
                    continue;
                }
                
                Queue<int[]> qu = new LinkedList<>();
                qu.add(new int[]{y, x});
                land[y][x] = 0;
                Set<Integer> connect = new HashSet<>();
                int amount = 0;
                
                while (!qu.isEmpty()) {
                    int[] pos = qu.poll();
                    int y1 = pos[0], x1 = pos[1];
                    amount++;
                    connect.add(x1);
                    
                    for (int[] direction : directions) {
                        int dy = direction[0], dx = direction[1];
                        int ny = y1 + dy, nx = x1 + dx;
                        if (nx >= 0 && nx < oilNum && ny >= 0 && ny < depth && land[ny][nx] == 1) {
                            qu.add(new int[]{ny, nx});
                            land[ny][nx] = 0;
                        }
                    }
                }
                
                for (int num : connect) {
                    oils[num] += amount;
                }
            }
        }
        
        return Arrays.stream(oils).max().orElse(0);
    }
}
